File: 9.Minimum-Spanning-Tree_Shortest_Path/근우/4.path-with-maximum-probability.py
## https://leetcode.com/problems/path-with-maximum-probability/

class Solution:
    def maxProbability(self, n: int, edges: List[List[int]], succProb: List[float], start: int, end: int) -> float:

        graph = defaultdict(dict)

        for edge, prob in zip(edges, succProb):
            graph[edge[0]][edge[1]] = prob
            graph[edge[1]][edge[0]] = prob

        queue = [[-1, start]]
        visited = set()

        while queue:
            prob, node = heapq.heappop(queue)

            if node == end:
                return -prob

            visited.add(node)

            for next_node, next_prob in graph[node].items():
                if next_node in visited:
                    continue

                heapq.heappush(queue, [prob * next_prob, next_node])

        return 0

File: 9.Minimum-Spanning-Tree_Shortest_Path/동규/4.path-with-maximum-probability.js
// https://leetcode.com/problems/path-with-maximum-probability/

var maxProbability = function (n, edges, succProb, start, end) {
  const probabilities = new Array(n).fill(0);
  probabilities[start] = 1;

  for (let i = 0; i < n; i += 1) {
    let betterPathFound = false;
    for (let j = 0; j < edges.length; j += 1) {
      const [from, to] = edges[j];
      if (!probabilities[from] && !probabilities[to]) continue;

      const edgeProb = succProb[j];
      const maxProb =
        Math.max(probabilities[from], probabilities[to]) * edgeProb;
      if (maxProb > probabilities[to]) {
        probabilities[to] = maxProb;
        betterPathFound = true;
      }
      if (maxProb > probabilities[from]) {
        probabilities[from] = maxProb;
        betterPathFound = true;
      }
    }
    if (!betterPathFound) break;
  }
  return probabilities[end];
};
